### ISSUE TITLE
projectile-find-file-in-directory does not work in TRAMP projects

### ISSUE THREAD

**Issue by codemac** [bbatsov/projectile#1275]
https://github.com/bbatsov/projectile/blob/a4b447d980b10cbb8d175d64e4305b4504c03d83/projectile.el#L3569

The resolution probably relates to expand-file-name being passed a tramp directory. I will post here any updates I find to fixing it.

**Comment by codemac** [comment]
Currently it attempts to open the file on the local FS, so /ssh:machine:/blah/blah ==> /blah/blah locally.

**Comment by codemac** [comment]
projectile-indexing-method 'native solves this, but takes a super long time to index, so this may be a deeper issue with how the alien method returns remote projects' name.

**Comment by codemac** [comment]
Yeah, so it seems the cached filenames need to be normalized when indexing between alien & native methods for tramp projects. The absolute paths the alien method produces shortcut the TRAMP default-directory. The `expand-file-name` method returns the name without the tramp default directory when handed an absolute filename.

**Comment by bbatsov** [comment]
@codemac Mind putting some PR together to tackle this. 

> Yeah, so it seems the cached filenames need to be normalized when indexing between alien & native methods for tramp projects.

Which cached filenames? Those in the project cache, right?

**Comment by codemac** [comment]
Sorry, @bbatsov yes these are the cached project filenames.

**Comment by codemac** [comment]
I stopped using TRAMP, and only use projectile locally due to other extensions issues with assuming local buffers.

I don't have time to implement this, but I do think this is still an issue for projectile's file cache.

**Comment by TrunovS** [comment]
> Currently it attempts to open the file on the local FS, so /ssh:machine:/blah/blah ==> /blah/blah locally.

Yes. Alien indexing method returned absolute file paths  which in turn failed `projectile-find-file` in Tramp session for me. Got that working with fix in the way `projectile-dir-files-alien` create list of files with relative paths instead of absolute paths. And it seams that fixed `find-file-in-directory` too.

**Comment by thornjad** [comment]
I use TRAMP every day and am entirely unable to use Projectile on any remote project. I'd really like to and don't want this issue to be closed as stale.

**Comment by TrunovS** [comment]
> I use TRAMP every day and am entirely unable to use Projectile on any remote project. I'd really like to and don't want this issue to be closed as stale.

Got that working with some combination of:
  - ```(add-to-list 'tramp-remote-path 'tramp-own-remote-path)```
 - ```(setq-default projectile-indexing-method 'alien)```
 - create empty .projectile in root of remote project path

**Comment by thornjad** [comment]
I hadn't tried `(add-to-list 'tramp-remote-path 'tramp-own-remote-path)` before, but it didn't seem to help. I have all three of those in place now, but any time I try to use `projectile-find-file` or any of its friends, it just prompts me to switch to a project.

**Comment by TrunovS** [comment]
> I hadn't tried `(add-to-list 'tramp-remote-path 'tramp-own-remote-path)` before, but it didn't seem to help. I have all three of those in place now, but any time I try to use `projectile-find-file` or any of its friends, it just prompts me to switch to a project.

Try this one in projectile.el:
```defcustom projectile-hg-command "hg files -0 ." ```

**Comment by thornjad** [comment]
Does anyone have any other suggestions? Most of my work is done on remote servers, and I'd really love to use projectile.

**Comment by adithyaov** [comment]
@thornjad 

I don't know exactly where you're stuck but this works for me.

You can use `(setq projectile-indexing-method 'native)` along with `(setq projectile-enable-caching t)`

Occasionally, I have to sync the cache. It is slow only the first time you run `projectile-find-files` because it needs to index the entire project.

If you're regularly deleting files, this probably won't work for you.

**Comment by thornjad** [comment]
@adithyaov your suggestion doesn't have any effect for me. Nothing has, I just can't get projectile to work with tramp at all.

**Comment by adithyaov** [comment]
If it works for me and does not work for you, it's probably because we are using different versions of projectile or emacs. Maybe try again after updating emacs to 27.* and projectile to its latest commit.

**Comment by thornjad** [comment]
My Emacs is built from the `master` branch (which currently reports as version 28.0.50) and I'm on the latest commit of Projectile already

**Comment by adithyaov** [comment]
If nothing works, you can always use `sshfs`.

**Comment by thornjad** [comment]
TRAMP is significantly more performant than SSHFS, at least for my work; I'd rather have Projectile work properly than change my workflow to suit it

**Comment by thornjad** [comment]
Be gone, stale tag, this is still an issue

**Comment by goldfita** [comment]
Same problem here (projectile-20220828.1421 on Windows). Changing projectile-indexing-method to native fixes the problem, but besides taking 20x longer to index, I need to use hybrid because without it, projectile doesn't use .gitignore.

**Comment by chy-causer** [comment]
Just to add a "what fixed it for me", my projectile projects over TRAMP were showing no files until I uninstalled fd-find on my desktop running emacs. I understand there is a `projectile-git-use-fd`, which I did not test. The remote server does not have fd installed, and I have no means of installing it there sadly.

**Comment by bbatsov** [comment]
Sadly there's no easy way to detect what tools are available remotely, that's why Projectile tries to use remotely whatever it found locally. You can open an issue about this - at the very least we can add something to the docs about this.

**Comment by bbatsov** [comment]
The alien indexing method returns project-relative file names now, and Projectile detects `fd` per-host, so the TRAMP case that broke here works as of 3.0 (see also #1501). The most common trigger was `fd` being installed locally but not on the remote.

Closing - reopen if something still misbehaves on a current version.
